databroker/glue.py

```diff
@@ -0,0 +1,15 @@
+from __future__ import absolute_import, division, print_function
+
+from glue.core.data import Data, Component
+
+
+def read_header(header):
+    out = []
+    for stream in header.stream_names:
+        result = Data(label="{stream}_{uid}".format(stream=stream, uid=header.start['uid']))
+        tbl = header.table(stream, fill=True)
+        for col in tbl:
+            result.add_component(Component(tbl[col].values), col)
```

Comment:
`.values` is often a surprisingly expensive call. Does `add_component` definite need a numpy array or can it manage with a `pandas.Series`? If it needs a numpy array, I'd suggest doing `.values` once in the outer loop.

```python
tbl = header.table(stream, fill=True)
arr = tbl.values
for i, col in enumerate(tbl.columns):
    result.add_component(Component(arr[i]), col))
```

Reply:
Thanks for the useful recommendation, @danielballan, I'll test it and will update the code.

Reply:
looks like it takes [numpy array](https://github.com/glue-viz/glue/blob/master/glue/core/component.py#L77). Nice catch.

Reply:
According to https://github.com/glue-viz/glue/blob/0e565f9e2b064bac6b6b1a10efd9ce028c45404e/glue/core/component.py#L43 it expects numpy arrays.

Reply:
`Component.__init__` looks like it will fail on anything but a literal numpy array, but it looks like `add_component` has some brains to delegate to various `Component` subclasses. Here's a place to start: https://github.com/glue-viz/glue/blob/f30c2d6e134e5525f5420d35920b74e0828c9bce/glue/core/data_factories/pandas.py#L109

Reply:
Yes, it fails when the `pandas.Series` type is used:
```py
Traceback (most recent call last):
  File "pandas\_libs\index.pyx", line 139, in pandas._libs.index.IndexEngine.get_loc
  File "pandas\_libs\hashtable_class_helper.pxi", line 811, in pandas._libs.hashtable.Int64HashTable.get_item
TypeError: an integer is required

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\utils\misc.py", line 58, in result
    return func(*args, **kwargs)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\app\qt\layer_tree_widget.py", line 120, in _do_action
    self.app.choose_new_data_viewer(data)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\app\qt\application.py", line 841, in choose_new_data_viewer
    return self.do(cmd)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\core\application_base.py", line 218, in do
    return self._cmds.do(command)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\core\command.py", line 127, in do
    result = cmd.do(self._session)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\core\command.py", line 217, in do
    v = session.application.new_data_viewer(self.viewer, self.data)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\utils\matplotlib.py", line 158, in wrapper
    result = func(*args, **kwargs)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\core\application_base.py", line 85, in new_data_viewer
    if data and not c.add_data(data):
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\utils\matplotlib.py", line 154, in wrapper
    return func(*args, **kwargs)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\viewers\common\qt\data_viewer_with_state.py", line 76, in add_data
    layer = self.get_data_layer_artist(data)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\utils\matplotlib.py", line 154, in wrapper
    return func(*args, **kwargs)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\viewers\common\qt\data_viewer_with_state.py", line 104, in get_data_layer_artist
    return self.get_layer_artist(self._data_artist_cls, layer=layer, layer_state=layer_state)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\utils\matplotlib.py", line 154, in wrapper
    return func(*args, **kwargs)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\viewers\matplotlib\qt\data_viewer.py", line 118, in get_layer_artist
    return cls(self.axes, self.state, layer=layer, layer_state=layer_state)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\viewers\scatter\layer_artist.py", line 51, in __init__
    layer_state=layer_state, layer=layer)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\viewers\matplotlib\layer_artist.py", line 27, in __init__
    layer=self.layer)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\viewers\scatter\state.py", line 235, in __init__
    self._on_layer_change()
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\viewers\scatter\state.py", line 253, in _on_layer_change
    self.cmap_att_helper.set_multiple_data([self.layer])
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\core\data_combo_helper.py", line 270, in set_multiple_data
    self.refresh()
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\core\data_combo_helper.py", line 302, in refresh
    if ((comp.numeric and self.numeric) or
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\core\component.py", line 104, in numeric
    return np.can_cast(self[(0,) * self.ndim].dtype, np.complex)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\glue\core\component.py", line 93, in __getitem__
    return self._data[key]
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\pandas\core\series.py", line 623, in __getitem__
    result = self.index.get_value(self, key)
  File "C:\Miniconda3\envs\bluesky-1.0b\lib\site-packages\pandas\core\indexes\base.py", line 2557, in get_value
    tz=getattr(series.dtype, 'tz', None))
  File "pandas\_libs\index.pyx", line 83, in pandas._libs.index.IndexEngine.get_value
  File "pandas\_libs\index.pyx", line 91, in pandas._libs.index.IndexEngine.get_value
  File "pandas\_libs\index.pyx", line 141, in pandas._libs.index.IndexEngine.get_loc
KeyError: (0,)
```
I'll look at the factories.

Reply:
9c705fa contains the suggested improvements, it's working fine within the GUI. Thanks for the hint, @danielballan, the code is now very similar to:
```py
    @qglue_parser(pd.DataFrame)
    def _parse_data_dataframe(data, label):
        label = label or 'Data'
        result = Data(label=label)
        for c in data.columns:
            result.add_component(data[c], str(c))
        return [result]
```